rclcpp/include/rclcpp/experimental/intra_process_manager.hpp

```diff
@@ -394,9 +398,13 @@ class IntraProcessManager
       }
       auto subscription_base = subscription_it->second.subscription.lock();
       if (subscription_base) {
-        auto subscription = std::static_pointer_cast<
+        auto subscription = std::dynamic_pointer_cast<
           rclcpp::experimental::SubscriptionIntraProcess<MessageT>
```

Comment:
@wjwwood shouldn't this be
```suggestion
          rclcpp::experimental::SubscriptionIntraProcess<MessageT, Alloc, Deleter>
```
implying publisher and subscription have the same allocator and deleter? Same above (we'll need extra template parameters).

Reply:
That might be the fix, I was gonna write a test cast first. I hadn't investigated the issue yet.

Reply:
I might need to take a break to work on something else, but later tonight I'll try this to see if it fixes it.

Reply:
Looks like this is the solution, see: 490cca76

The tests pass now. 🎉